rfcs/2020-12-08-5448-dynamic-sampling.md

```diff
@@ -0,0 +1,351 @@
+sampled (i.e. ending up in any bucket). I propose retaining the exclusion option but allowing users
+to specifiy criteria via `check_fields` *or Remap* (only `check_fields` is currently available).
+
+### Bucket behavior
+
+With events separated into buckets you can begin specifying *how* events in those buckets are
+sampled. The sections below propose per-bucket parameters. All of these parameters can apply to
+named *or* dynamic buckets. For named buckets, these parameters are set with the bucket definition;
+for dynamic buckets, the parameters would be set at the transform level and apply to all created
+buckets.
+
+#### `rate`
+
+The base sampling rate for the bucket. If only this parameter is specified, the bucket is constantly
+sampled at a rate of 1/N.
+
+#### `sensitivity`
+
+If this parameter is specified, that means that a dynamic sampling algorithm is applied to the
+bucket. This determines how Vector adjusts the sample rate, using `rate` as the baseline, in light
+of changes in event frequency. A sensitivity close to 0 would mean only small adjustments whereas a
+sensitivity close to 1 would mean more dramatic adjustments.
+
+> [Adaptive Request Concurrency][arc]'s `decrease_ratio` parameter serves as a rough analogy.
+
+The parameters below only make sense if `sensitivity` is defined.
+
+#### `max_rate` / `min_rate`
+
+Optional parameters to keep the sample rate within hard limits.
+
+#### `min_event_threshold`
+
+The minimum number of events needed, within the time window, to begin adjusting the sample rate.
+Below this threshold, `rate` applies. Defaults to 0, meaning no threshold.
+
+#### `window_secs`
+
+This specifies the length of the time window, in seconds, during which the sampling rate is
+calculated. No dynamic rate can be calculated without a window. I propose 15 seconds as a default
+but this can be determined during the development process.
+
+### Metrics
+
+I propose gathering the following additional metrics from the `sampler` transform:
+
+* A gauge for the total number of current buckets
+* A counter for the total number of buckets created
+
+## Doc-level Proposal
+
+More robust dynamic sampling capabilities would require the following documentation changes:
+
+* Either update the existing `sampler` transform docs to include the new configuration parameters
+  plus some additional explanation.
+* A new dynamic sampling guide that walks through specific sampling use cases.
+* A new "Under the Hood" page à la the one we provide for Adaptive Request Concurrency that explains
+  the feature in more detail, with accompanying diagrams
+
+## Prior Art
+
+Other systems in the observability space do provide dynamic sampling capabilities. Their feature
+sets and user interfaces can provide inspiration for dynamic sampling in Vector (and have influenced
+this RFC). In terms of implementation, however, I'm unaware of a system built in Rust that provides
```

Comment:
We've often linked to prior art of other applications (like fluentbit, logstash, etc.) regardless of their implementation language. It is still useful to compare their implementation and the configuration options they chose to expose.

Would you be able to add some you are aware of?

I did find something kinda analogous in logstash: https://www.elastic.co/guide/en/logstash/current/plugins-filters-drop.html